File: soyeong/programmers-2/큰_수_만들기.py
def solution(number, k):
    # 숫자 제거하고 나머지 숫자는 위치가 변하지 않은 상태에서 가장 큰 숫자를 찾는것
    answer = []

    for i in number:
        # 1) 빠져야할 숫자가 남아있으면
        # 2) answer에 숫자가 있다면
        # 3) answer의 마지막 값이 i보다 작다면
        while k > 0 and answer and answer[-1] < i:
            # 숫자를 pop해 제거하고
            answer.pop()
            # 숫자를 제거했으니 빼야할 숫자 개수를 줄인다.
            k -= 1
        # 그 외에는 answer 값에 추가
        answer.append(i)

    # 만약 위에서 k를 모두 제거했다면 k=0 이겠지만, 모두 같은 수등의 예외가 있다면 k는 하나도 제거되지 않은채 여기까지 도달함
    # 따라서 제거되지 않은 숫자 나머지를 답변 전에 answer 길이에서부터 k를 빼 제거하여 큰수를 생성
    # len(answer)-k를 하는 이유는 그냥 -k를 하게되면 0 경우 값이 출력이 안됨
    return ''.join(answer[:len(answer) - k])

File: kyulee/programmers-2/큰수만들기.java

public class 큰수만들기 {
class Solution {
    public String solution(String number, int k) {
        int len = number.length();
        int remain = len - k; // 남은 숫자의 개수
        int start = 0; // 검사할 시작 위치

        StringBuilder answer = new StringBuilder();

        // 남은 숫자의 개수만큼 반복
        while (remain > 0) {
            int maxIdx = start; // 현재 범위에서 가장 큰 숫자의 인덱스
            char maxChar = number.charAt(start); // 현재 범위에서 가장 큰 숫자

            // 현재 범위 내에서 가장 큰 숫자 찾기
            for (int i = start + 1; i <= len - remain; i++) {
                char currentChar = number.charAt(i);
                if (currentChar > maxChar) {
                    maxChar = currentChar;
                    maxIdx = i;
                }
            }

            answer.append(maxChar); // 가장 큰 숫자 추가
            start = maxIdx + 1; // 다음 범위의 시작 위치 설정
            remain--; // 한 숫자를 추가했으므로 남은 숫자의 개수 감소
        }

        return answer.toString();
        }
    }
}
